_content/devices/bytes.md

```diff
@@ -158,38 +166,73 @@ int myVal = ((int)(payload[0]) << 8)
 payload[2] = (byte) ((lng & 0x0000FF00) >> 8  );
 payload[3] = (byte) ((lng & 0X000000FF)       );
 ```
 
+For Arduino you will often see that the masks such as `0x00FF0000`, and the type casts to `(byte)`, are not used. That works when `payload` is defined as an array of `byte` as that will really only fit 8 bits, where only the 8 least significant bits are stored and all others are simply discarded. But it never hurts to be explicit, and other platforms might surely need the masks and casts.
+
 Decode (payload functions):
 
 ```js
-decoded.myVal = ((long)(bytes[0]) << 24)
-              + ((long)(bytes[1]) << 16)
-              + ((long)(bytes[2]) << 8)
-              + ((long)(bytes[3]));
+// For unsigned 32 bits number
+decoded.myVal = bytes[0] * 0x1000000
+              + bytes[1] * 0x10000
+              + bytes[2] * 0x100
+              + bytes[3];
 ```
 
+Note that in JavaScript, for _unsigned_ 32 bits numbers, we cannot use bitwise operators such as shifting or the bitwise OR, as those always assume a two's complement representation which might yield negative numbers.
+
 ## How to send negative numbers?
 
-To tell the difference between -100 and 100 you will need a [signed](https://en.wikipedia.org/wiki/Signed_number_representations) data type. These set the highest (left-most) bit to `1` to indicate it's a negative number. This does mean that for example in a word only 15 of the 16 bits are available for the actual number, limiting the range from 65536 to 32768.
+To tell the difference between -100 and 100 you will need a [signed](https://en.wikipedia.org/wiki/Signed_number_representations) data type. These limit the maximum values by half: where an unsigned word (16 bits) can hold values from 0 through 65536, a signed word ranges from -32768 through 32767. On many platforms, including Arduino, signed integers use a [two's complement representation](https://en.wikipedia.org/wiki/Two%27s_complement) in memory, in which negative numbers might seem to have little in common with their positive counterpart: where decimal 2348 is represented as hexadecimal 0x092C (or 0x0000092C for 32 bits), the negative -2348 would be 0xF6D4 (or 0xFFFFF6D4 for 32 bits).
 
 ### Index, round and shift
 
 The data types we used so far are all signed, which means all of the tricks work just as well for negative values. Just be aware of the maximum value.
 
+When decoding negative numbers using JavaScript, beware that JavaScript's bitwise operators always expect 32 bits, in a two's complement representation. When handling fewer bits, [sign extension](https://en.wikipedia.org/wiki/Sign_extension) is needed to support negative values, yielding code such as:
+
+```js
+// Sign-extend to support negative numbers: shifting
+// 24 bits leftwards, followed by shifting 16 bits
+// to the right, extends a "two's complement" negative
+// value such as 0xF6D4 into 0xFFFFF6D4.
+decoded.myVal = (bytes[0] << 24 >> 16)
+              + bytes[1];
```

Comment:
Also consistency with `+` and `|` (I prefer the latter)

Reply:
Again I agree with Johan here.